streams/src/main/java/org/apache/kafka/streams/processor/internals/StreamTask.java

```diff
@@ -875,4 +917,37 @@ private void initializeTransactions() {
             );
         }
     }
+
+    // visible for testing
+    String encodeTimestamp(final long partitionTime) {
```

Comment:
I would put methods to write and read record metadata in their own classes. Those classes would be kind of SerDes for metadata. Such SerDes would make the code better testable and separates the concerns of a task and reading and writing record metadata which are completely independent. It does not need to be done in this PR. I just wanted to mention it.

Reply:
Yeah, that would probably be a good idea in the future.